pkg/collector/corechecks/ebpf/probe/ebpfcheck/map_prog_helper.go

```diff
@@ -0,0 +1,201 @@
+					Name: "map",
+					Type: u64Btf,
+				},
+				{
+					Name: "key",
+					Type: u64Btf,
+				},
+				{
+					Name: "value",
+					Type: u64Btf,
+				},
+				{
+					Name: "ctx",
+					Type: u64Btf,
+				},
+			},
+		},
+	}
+
+	return &mapProgHelperCache{
+		cache: make(map[ebpf.MapID]helperProgData),
+
+		entryBtf:    entryBtf,
+		callbackBtf: callbackBtf,
+	}
+}
+
+func (c *mapProgHelperCache) newCachedProgramForMap(mp *ebpf.Map, mapid ebpf.MapID) (*ebpf.Program, error) {
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+
+	if data, ok := c.cache[mapid]; ok {
+		return data.prog, nil
+	}
+
+	data, err := c.newHelperProgramForFd(mp.FD())
+	if err != nil {
+		return nil, err
+	}
+
+	c.cache[mapid] = data
+	return data.prog, nil
+}
+
+type helperProgData struct {
+	id   ebpf.ProgramID
+	prog *ebpf.Program
+}
+
+func (c *mapProgHelperCache) newHelperProgramForFd(fd int) (helperProgData, error) {
+	/*
+		equivalent to the following C code, based on the fact that `for_each_map_elem`
+		returns the amount of entries visited, so visiting all the entries ensures we
+		get the amount of entries in the map.
+
+		int callback(struct bpf_map* map, const void* key, void* value, void * ctx) {
+			return 0;
+		}
+
+		int entry() {
+			return bpf_for_each_map_elem(fd, callback, NULL, 0);
+		}
+	*/
+
+	spec := &ebpf.ProgramSpec{
+		Type: ebpf.SocketFilter,
+		Instructions: asm.Instructions{
+			// entry
+			btf.WithFuncMetadata(
+				asm.LoadMapPtr(asm.R1, fd), // map fd
+				c.entryBtf,
+			),
+			asm.Instruction{
+				OpCode:   asm.LoadImmOp(asm.DWord),
+				Src:      asm.PseudoFunc,
+				Dst:      asm.R2,
+				Constant: -1,
+			}.WithReference("callback"),
+			asm.LoadImm(asm.R3, 0, asm.DWord), // callback ctx
+			asm.LoadImm(asm.R4, 0, asm.DWord), // flags
+			asm.FnForEachMapElem.Call(),
+
+			asm.Instruction{
+				OpCode: asm.Exit.Op(asm.ImmSource),
+			},
+
+			// callback
+			btf.WithFuncMetadata(
+				asm.LoadImm(asm.R0, 0, asm.DWord).WithSymbol("callback"),
+				c.callbackBtf,
+			),
+			asm.Instruction{
+				OpCode: asm.Exit.Op(asm.ImmSource),
+			},
+		},
+		License: "GPL",
+	}
+
+	prog, err := ebpf.NewProgramWithOptions(spec, ebpf.ProgramOptions{
+		LogDisabled: true,
+	})
+	if err != nil {
+		return helperProgData{}, err
+	}
+
+	var progid ebpf.ProgramID
+	info, err := prog.Info()
+	if err == nil {
+		if id, ok := info.ID(); ok {
+			ddebpf.AddIgnoredProgramID(id)
+			progid = id
+		}
+	}
+
+	return helperProgData{
+		id:   progid,
+		prog: prog,
+	}, nil
+}
+
+func (c *mapProgHelperCache) Close() {
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+
+	for _, data := range c.cache {
+		if err := data.prog.Close(); err != nil {
+			log.Warnf("failed to close helper program: %v", err)
+		}
+		ddebpf.RemoveIgnoredProgramID(data.id)
+	}
+	clear(c.cache)
+}
+
+func (c *mapProgHelperCache) closeUnusedPrograms() {
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+
+	for id, data := range c.cache {
+		_, err := ddebpf.GetMapNameFromMapID(uint32(id))
```

Comment:
This doesn't necessarily mean the map doesn't exist anymore, just that we don't have a full name for it.